velox/docs/develop/aggregate-functions.rst

```diff
@@ -109,25 +109,323 @@ accordingly.  Data is stored in the following order:
 .. image:: images/aggregation-layout.png
   :width: 600
 
-Aggregate class
----------------
-
-To add an aggregate function,
+To add an aggregate function, there are two options: implementing it as a
+simple function or as a vector function. The simple-function interface allows
+the author to write methods that process input data one-row-at-a-time and not
+handle input vector encodings themselves. However, the simple-function
+interface currently has certain limitations, such as not allowing for advanced
+performance optimization on constant inputs. Aggregation functions that
+require such functionalities can be implemented through the vector-function
+interface. With the vector-function interface, the author writes methods that
+process one-vector-at-a-time and handles input vector encodings by themselves.
 
 * Prepare:
     * Figure out what are the input, intermediate and final types.
     * Figure out what are partial and final calculations.
     * Design the accumulator. Make sure the same accumulator can accept both raw
       inputs and intermediate results.
-    * Create a new class that extends velox::exec::Aggregate base class
+    * If implementing a simple function, create a class for the function according
+      to instructions below; If implementing a vector function,
+      create a new class that extends velox::exec::Aggregate base class
       (see velox/exec/Aggregate.h) and implement virtual methods.
 * Register the new function using exec::registerAggregateFunction(...).
 * Add tests.
 * Write documentation.
 
-Accumulator size
+Simple Functions
+----------------
+
+This section describes the main concepts and the interface of simple
+Aggregation functions. Examples of aggregation functions implemented through
+the simple-function interface can be found at velox/exec/tests/SimpleAverageAggregate.cpp
+and velox/exec/tests/SimpleArrayAggAggregate.cpp.
+
+A simple aggregation function is implemented as a class as the following.
+
+.. code-block:: c++
+
+  // array_agg(T) -> array(T) -> array(T)
+  class ArrayAggAggregate {
+   public:
+    // Type(s) of input vector(s) wrapped in Row.
+    using InputType = Row<Generic<T1>>;
+    using IntermediateType = Array<Generic<T1>>;
+    using OutputType = Array<Generic<T1>>;
+
+    // Optional. Default is true.
+    static constexpr bool default_null_behavior_ = false;
+
+    // Optional.
+    static bool toIntermediate(
+      exec::out_type<Array<Generic<T1>>>& out,
+      exec::optional_arg_type<Generic<T1>> in);
+
+    struct AccumulatorType { ... };
+  };
+
+The author declares the function's input type, intermediate type, and output
+type in the simple aggregation function class. The input type must be the
+function's argument type(s) wrapped in a Row<> even if the function only takes
```

Comment:
This is needed because we don't know how many input types an aggregation function have, so we use this way to allow author to define them as one type and then parse the fields of Row<>. I added a sentence "This is needed for the SimpleAggregateAdapter to parse input types for arbitrary aggregation functions properly".